be/src/exec/pipeline/exchange/sink_buffer.h

```diff
@@ -34,171 +32,28 @@ struct TransmitChunkInfo {
-        size_t buffer_size = 0;
-        for (auto& [_, buffer] : _buffers) {
-            buffer_size += buffer.size();
-        }
-        return buffer_size > max_buffer_size;
-    }
-
-    bool is_finished() const {
-        if (!_is_finishing) {
-            return false;
-        }
-
-        return _num_sending_rpc == 0 && _num_in_flight_rpc == 0;
-    }
+    SinkBuffer(RuntimeState* state, const std::vector<TPlanFragmentDestination>& destinations, size_t num_sinkers);
+    ~SinkBuffer();
+
+    void add_request(const TransmitChunkInfo& request);
+    bool is_full() const;
+    bool is_finished() const;
 
     // When all the ExchangeSinkOperator shared this SinkBuffer are cancelled,
     // the rest chunk request and EOS request needn't be sent anymore.
-    void cancel_one_sinker() {
-        if (--_num_uncancelled_sinkers == 0) {
-            _is_finishing = true;
-        }
-    }
+    void cancel_one_sinker();
 
 private:
-    void _try_to_send_rpc(const TUniqueId& instance_id) {
-        DeferOp defer([this]() { --_num_sending_rpc; });
-        ++_num_sending_rpc;
-
-        if (_is_finishing) {
-            return;
-        }
-
-        for (;;) {
-            auto& buffer = _buffers[instance_id.lo];
-            auto* closure = _closures[instance_id.lo];
-            if (buffer.empty() || closure->is_in_flight()) {
-                return;
-            }
-
-            TransmitChunkInfo request = buffer.front();
-            buffer.pop();
-
-            if (request.params->eos()) {
-                if (--_num_remaining_eos == 0) {
-                    _is_finishing = true;
-                }
-                // Only the last eos is sent to ExchangeSourceOperator. it must be guaranteed that
-                // eos is the last packet to send to finish the input stream of the corresponding of
-                // ExchangeSourceOperator and eos is sent exactly-once.
-                if (--_num_sinkers[instance_id.lo] > 0) {
-                    if (request.params->chunks_size() == 0) {
-                        request.params->release_finst_id();
-                        continue;
-                    } else {
-                        request.params->set_eos(false);
-                    }
-                }
-            }
-
-            request.params->set_allocated_finst_id(&_instance_id2finst_id[instance_id.lo]);
-            request.params->set_sequence(_request_seqs[instance_id.lo]++);
-
-            ++_num_in_flight_rpc;
-
-            closure->ref();
-            closure->cntl.Reset();
-            closure->cntl.set_timeout_ms(_brpc_timeout_ms);
-            closure->cntl.request_attachment().append(request.attachment);
-            request.brpc_stub->transmit_chunk(&closure->cntl, request.params.get(), &closure->result, closure);
-
-            request.params->release_finst_id();
-            return;
-        }
-    }
+    void _process_send_window(const TUniqueId& instance_id, const int64_t sequence);
```

Comment:
add some comment to help us to understand the semantics.

Reply:
ok